backend/app/lib/background_job_queue.rb

```diff
@@ -3,68 +3,75 @@
 
 require 'thread'
 require 'atomic'
+
 require_relative 'job_runner'
-require_relative 'find_and_replace_runner'
-require_relative 'print_to_pdf_runner'
-require_relative 'reports_runner'
-require_relative 'batch_import_runner'
-require_relative 'container_conversion_runner'
+
+# load job runners
+Dir.glob(File.join(File.dirname(__FILE__), "job_runners", "*.rb")).each do |file|
+  require file
+end
+
+# and also from plugins
+ASUtils.find_local_directories('backend').each do |prefix|
+  Dir.glob(File.join(prefix, "job_runners", "*.rb")).each do |file|
+    require File.absolute_path(file)
+  end
+end
+
 
 class BackgroundJobQueue
 
   JOB_TIMEOUT_SECONDS = AppConfig[:job_timeout_seconds].to_i
 
-  def find_stale_job
-    DB.open do |db|
-      stale_job = Job.any_repo.
-        filter(:status => "running").
-        where {
-        system_mtime <= (Time.now - JOB_TIMEOUT_SECONDS)
-      }.first 
-
-      if stale_job
-        begin
-          stale_job.time_started = Time.now
-          stale_job.save
-          return stale_job
-        rescue
-          # If we failed to save the job, another thread must have grabbed it
-          # first.
-          nil
-        end
+  def get_next_job
+    # First cancel any jobs that are in a running state but which haven't
+    # been touched by their watchdog for a time greater than the configured timeout.
+    # This shouldn't really happen, but his replaces the concept of a stale job
+    # used in an earlier implementation that was problematic because it could end up
+    # calling the #run method on a job more than once.
+    begin
+      Job.running_jobs_untouched_since(Time.now - JOB_TIMEOUT_SECONDS).each do |job|
+        job.finish!(:canceled)
       end
+    rescue Sequel::NoExistingObject
+      Log.debug("Another thread cancelled unwatched job #{job.id}, nothing to do on #{Thread.current[:name]}")
+    rescue => e
+      Log.error("Error trying to cancel unwatched jobs on #{Thread.current[:name]}: #{e.class} #{$!} #{$@}")
     end
-  end
 
+    DB.open do |db|
+      Job.queued_jobs.each do |job|
+        runner = JobRunner.registered_runner_for(job.type)
 
-  def find_queued_job
-    while true
-      DB.open do |db|
+        begin
+          unless runner
+            Log.error("No runner registered for #{job.type} job #{job.id}! " +
+                      "Marking as canceled on #{Thread.current[:name]}")
 
-        job = Job.any_repo.
-          filter(:status => "queued").
-          order(:time_submitted).first
+            job.finish!(:canceled)
+            next
+          end
 
-        return unless job
+          if !runner.run_concurrently and Job.any_running?(job.type)
```

Comment:
I used `and` because `&&` was binding more tightly than the `!` and I thought `and` was less noisy than brackets